### ISSUE TITLE
bench_mlp rejects BENCH_TUNE with BENCH_PRECISION, so bf16 tensor-core reachability cannot be measured — blocks the HIP leg's step-0 diagnostic

### ISSUE THREAD

**Issue by lukstafi** [ahrefs/ocannl#529]
Found by the gh-ocannl-476 CUDA leg, and it will block the HIP leg's step-0 diagnostic.

## Symptom

`benchmarks/runners/ocannl/bench_mlp.ml:48-49` rejects the combination outright:

```ocaml
if Option.is_some mp_prec && tune then
  failwith "bench_mlp: BENCH_TUNE with BENCH_PRECISION is not supported";
```

So the mixed-precision legs (`orchestrate.py --precision bf16 f16`) are **untuned-only**, and there
is no way to run an autotune search over a reduced-precision graph through the benchmark harness.

## Why it matters for the tensor-core arc specifically

Whether a tensorized candidate is even *seeded* depends on the operand storage precision, via
`Autotune.mma_input_formats_of_prec` -> `mma_tile_for_precisions` against the backend's
`mma_format_tiles`. That makes bf16 the one configuration whose native 16x16x16 tile is advertised
unconditionally:

- **CUDA**: at uniform f32 there is no `Mma_f32` entry at all, so zero mma candidates are seeded;
  the question is only askable via `tf32_matmuls=true`. bf16 would be the route that does not
  depend on the numerics policy — and it cannot be tuned.
- **HIP**: RDNA3/3.5 WMMA has only f16xf16 and bf16xbf16 at 16x16x16 and **no f32-input shape**
  (`arrayjit/lib/hip_backend.ml:531`), so uniform f32 stays scalar and there is no tf32 arm to fall
  back on. **bf16 is that backend's only tensor-core route, and this guard makes it unmeasurable.**
  gh-ocannl-476's handoff asks the HIP leg to run the same candidate-level step-0 diagnostic that
  the macOS and CUDA legs ran; as things stand it cannot be run there at all via `BENCH_TUNE`.

A second, smaller consequence: every bf16/f16 number currently in the reports is by construction a
*storage/bandwidth* result with no tensorized rendering behind it. That is worth being explicit
about when reading them — on CUDA bf16 is a slight win (`mlp_small/cuda/bf16` 0.103 vs 0.112 ms
f32), and none of it is tensor cores.

## What would resolve it

Either lift the restriction (tune the reduced-precision graph directly), or add a narrower escape
hatch for the diagnostic case — a mode that runs the seeding/compile phase and reports candidate
outcomes without needing the full timed search to be meaningful. The latter is enough for step-0
style questions ("are tensorized candidates proposed, and do they compile?") and avoids whatever
made the combination unsafe to time in the first place.

Worth recording in the commit/issue why the guard exists, which is not obvious from the code — if
it is about loss-scaling state interacting with repeated timed replays, that would only rule out
f16, not bf16 (bf16 carries no gate).

**Comment by lukstafi** [comment]
## The HIP leg confirms this, and it is worse than "blocks the step-0 diagnostic"

I filed #534 for this from the HIP leg before spotting this issue — **closing #534 as a duplicate**; recording its HIP-specific content here.

On gfx1151 the guard is not merely an obstacle to the diagnostic, it makes the measurement **structurally impossible**:

- RDNA3/3.5 WMMA has **no f32-input shape at all** (`hip_backend.ml:531`), unlike CUDA where the tf32 arm (`Numerics.tf32_matmuls`, gh-478) provides an f32-storage route independent of the reduced-precision recipe. So on HIP the f32 cells seed *zero* mma candidates — not "fewer", zero.
- Every rocWMMA route therefore runs through `BENCH_PRECISION`: f16xf16 -> f32 (the flagship, and the combination verified on gfx1151 via `schedule_mma_matmul`), f16xf16 -> f16, bf16xbf16 -> f32, bf16xbf16 -> bf16.
- With this guard in place, **no tuned HIP benchmark cell can contain a rocWMMA intrinsic, at any precision, regardless of scheduling state.** The HIP column cannot deliver the gh-152/155 payoff measurement no matter how much tuning time is spent on it.

Since CUDA has the tf32 escape hatch and HIP does not, this issue is the *sole* blocker for the HIP column while it is one of two for CUDA — worth weighting when it is scheduled.

Confirmed empirically: `BENCH_FIXTURE=fixtures/mlp_wide.safetensors BENCH_TUNE=1 BENCH_PRECISION=bf16 bench_mlp.exe --ocannl_backend=hip` dies immediately on the `failwith`. The HIP leg therefore ran its step-0 diagnostic on the f32 arm instead, which still established that #521 generalizes (all GPU sketch candidates fail candidate compile before being timed) but could say nothing about mma reachability.

Context: `benchmarks/report-hip.md` (Findings -> "Tensor cores on HIP: still zero, and structurally unreachable"), lukstafi/ocannl-staging#244.

**Comment by lukstafi** [comment]
## Restriction lifted (lukstafi/ocannl-staging#256)

Checked first whether it still stood as described after #250's rework — it did, unchanged at
`bench_mlp.ml:77-78`. On the "why does the guard exist" question: it was introduced in the same
commit as `BENCH_PRECISION` itself (`ea610155`, gh-ocannl-492 task 4), records no reason, and
nothing in the tuned path is precision-specific. It reads as "not exercised", not as a known
hazard.

**Every leg now tunes.** The f32, bf16 and f16-static legs already build a single fused `step_comp`
and needed only the guard removed. The dynamic-loss-scaling legs keep their step *shape* — the gate
is what those legs measure — so only the gradient (`HostGate`) or fused (`GATE_INTERVAL`) routine is
tuned, and the small optimizer routine keeps its plain compile, taken from the tuned context so the
lineage's compile order is unchanged. Tuning already runs candidates on a scratch lineage
(`timing_ctx`), so the repeated executions touch neither the benchmark's weights nor the host-side
scaler state.

**Verified on `cc`** across `f32`, `bf16`, `f16`, `f16-static` and `f16-gated2`: all five search
both placement arms and stay on their untuned loss trajectories. (bf16 tuned reports
`losses:[0.635861516, 0.57146275]` against untuned `[0.635861516, 0.571274459]` — the step-1
divergence is the materialize-all arm changing recompute rounding under bf16, which is expected.)

**The payoff still needs the HIP machine.** What this unblocks is the question, not the answer:
whether gfx1151 seeds a tensorized candidate at all under bf16, which is the only tensor-core route
there — RDNA3/3.5 WMMA has no f32-input shape, so uniform f32 stays scalar and there is no tf32 arm
to fall back on. The step-0 diagnostic from gh-ocannl-476's handoff is now runnable there:

```bash
cd benchmarks && BENCH_FIXTURE=fixtures/mlp_small.safetensors BENCH_TUNE=1 BENCH_PRECISION=bf16 \
  BENCH_TUNE_REPORT=1 ../_build/default/benchmarks/runners/ocannl/bench_mlp.exe \
  --ocannl_backend=hip --ocannl_autotune_log=true
```

`BENCH_TUNE_REPORT=1` prints each arm's `mma_candidates` / `mma_timed` counts and the decline census
keys, which is exactly the "were tensorized candidates proposed, and did they compile" answer —
without needing the timed search to be meaningful.

**One gap left deliberately.** `orchestrate.py` still treats `tuned` and `bf16` as mutually
exclusive *variants* (`BENCH_TUNE="1" if variant == "tuned"`, `BENCH_PRECISION=variant if variant in
("bf16","f16")`), so a tuned reduced-precision *matrix cell* is not expressible through the
orchestrator — only through a direct `bench_mlp` invocation as above. That is enough for the
diagnostic this issue is blocking, and adding a combined variant would touch the variant naming, the
`SKIP_CELLS` keys and the report grouping, so I left it out of scope. Worth filing separately if the
tuned bf16/f16 cells should become part of the reported matrix.

Leaving this open until the HIP run confirms the payoff.

**Comment by lukstafi** [comment]
## bf16 tensor-core reachability on HIP — measured (gfx1151)

With the `BENCH_TUNE` × `BENCH_PRECISION` guard removed, this is now answerable on the HIP column for the first time. Machine: Radeon 8060S (gfx1151, 20 CUs), ROCm 7.14, WSL2. Staging at `081e70e2` (#258 merged), autotune cache wiped before every run.

### The instrument needed a fix first

`BENCH_TUNE_REPORT=1` did **not** print `mma_candidates` / `mma_timed` — the fields exist on `Autotune.report`, but `bench_mlp`'s reporter never emitted them, so the arms were unreadable. One-line fix included in the staging PR; all numbers below come from the fixed reporter.

### Result: bf16 is the only route that seeds anything, and nothing reaches timing

`mlp_small`, tuned, both placement arms:

| precision | arm A (default placements) | arm B (materialize-all) |
|---|---|---|
| f32  | 0 seeded / 0 timed | 0 seeded / 0 timed |
| bf16 | 0 seeded / 0 timed | **28 seeded / 0 timed** |

The f32 legs confirm the structural claim: `hip_backend.ml` advertises `mma_format_tiles` of f16 and bf16 at 16×16×16 only, no f32 entry, so an f32 site seeds **zero** mma candidates and there is no tf32 escape hatch. bf16 is the only tensor-core route on RDNA3/3.5, exactly as expected.

So bf16 does unblock **seeding** — 28 candidates where f32 had none. But `mma_timed = 0`: not one tensorized pipeline was measured. Against the CUDA control (66 seeded / 32 timed / crowned `Tensorize` = 2 per arm), the HIP column has **not** yet reached the state #521's fix delivered on CUDA.

### Where all 28 die — accounts for every candidate

| blocker | count |
|---|---|
| `Schedule.Tensorize`: not a perfectly nested i×j×k micro-kernel | 12 |
| companion coverage (#521): companion nest `max_logits` aligned chain 64 vs accumulation 64×2 | 6 |
| companion coverage (#521): cross-nest race analysis bails, 98–99 nests cannot be given aligned geometry | 6 |
| `Schedule.Tensorize`: operands not indexed `[…,i,k]`/`[…,k,j]` (`n14_cast`, `n32_relu`) | 4 |

12 + 6 + 6 + 4 = 28. The two companion-coverage families (12) are the ones #258 types — they are recorded as `(Unsupported_key autotune_sketch_companion_coverage)` declines and the search continues rather than aborting. The other 16 are `Schedule.Tensorize` structural rejections, a different arc from #521's `validate_parallel` work.

Notably **no** "Stage: source is written in the routine" failures appeared here.

### Run

```
BENCH_FIXTURE=fixtures/mlp_small.safetensors BENCH_TUNE=1 BENCH_PRECISION=bf16 \
BENCH_TUNE_REPORT=1 ./bench_mlp.exe --ocannl_backend=hip --ocannl_autotune_log=true
```
Completed cleanly in ~30 s compile (`compile_s` 30.5 first run); loss trajectory matches the f32 leg to 3 s.f. No abort — #258 held at stock config, so the amendment's `--ocannl_strict_failure_classification=false` workaround was not needed.

### Not covered

The `lenet` leg is **not** measurable: it is a `model: conv` workload, so it runs through `bench_conv`, which supports neither `BENCH_PRECISION` nor `BENCH_TUNE_REPORT`. Extending it is a separate piece of work — the mixed-precision recipe in `bench_mlp` (loss scaling, gating) is not a small port. `mlp_wide` and `cifar_conv` tuned were avoided per the #532 dispatch hazard.

**Comment by lukstafi** [comment]
Closing: the restriction is lifted and the measurement it blocked has been taken. The answer is negative, and its successor is #540.

**The guard is gone** (lukstafi/ocannl-staging#256). It was introduced in the same commit as `BENCH_PRECISION` itself, recorded no reason, and nothing in the tuned path was precision-specific — it read as "not exercised", not as a known hazard. Every leg now tunes; the dynamic-loss-scaling legs keep their step *shape* since the gate is what they measure, so only the gradient or fused routine is tuned.

**The measurement, on gfx1151:** bf16 makes tensor-core candidates expressible for the first time — **28 seeded against 0 at f32**, which is the structural point this issue was about (RDNA3/3.5 WMMA has no f32-input shape, so f32 seeds nothing at all, and unlike CUDA there is no tf32 route). But **`mma_timed` is still 0.** The blockers are 12 `Tensorize` perfect-nest and 4 operand-indexing declines — a different subsystem from #521's coverage work, now tracked as **#540**.

So this issue's question — *can the HIP column's tensor-core reachability be measured at all?* — is answered yes. The payoff question moves to #540.

**One gap left deliberately**, now #539 (landed, lukstafi/ocannl-staging#259): `orchestrate.py` modelled tuning and precision as one variant axis, so tuned reduced-precision *matrix cells* were not expressible even after this fix — only direct `bench_mlp` invocations. That is what #538's re-measurement needs.
